### ISSUE TITLE
question about GRPO code

### ISSUE THREAD

**Issue by Hongbo-Jin** [tulerfeng/Video-R1#87]
在你的代码中关于GRPO的实现，你没有保存上一个step中的per token logps，而是直接减去了本轮的per token logps，这会导致importance ratio全部为1，进而导致所有advantage相加为0，对应部分的loss也全部为0，loss完全由kl散度贡献，整个RL训练过程将毫无意义。
你的代码实现如下：
        mean_grouped_rewards = mean_grouped_rewards.repeat_interleave(self.num_generations, dim=0)
        std_grouped_rewards = std_grouped_rewards.repeat_interleave(self.num_generations, dim=0)
        advantages = (rewards - mean_grouped_rewards) / (std_grouped_rewards + 1e-4)

        per_token_loss = torch.exp(per_token_logps - per_token_logps.detach()) * advantages.unsqueeze(1)
        per_token_loss = -(per_token_loss - self.beta * per_token_kl)
        loss = ((per_token_loss * completion_mask).sum(dim=1) / completion_mask.sum(dim=1)).mean()

请问能否解释下这部分，谢谢

**Comment by Hongbo-Jin** [comment]
感谢回复，请问advantage不在loss中的话，对应的策略梯度是怎么反向传播的，从哪里传过来呢







---- Replied Message ----
| From | Kaituo ***@***.***> |
| Date | 9/24/2025 17:40 |
| To | ***@***.***> |
| Cc | ***@***.***>,
***@***.***> |
| Subject | Re: [tulerfeng/Video-R1] question about GRPO code (Issue #87) |
tulerfeng left a comment (tulerfeng/Video-R1#87)

您好，grpo advantage方面和为0是正常的，可以参考这个issue：#50

另外，importance ratio 为1 是因为我们这个是on-policy （即rollout一次，模型更新一次， $\pi = \pi_{old}$），而不是类似dapo的off-policy更新

这些实现都和之前的大部分grpo工作保持一致
